tests/test_builtin_casters.py

```diff
@@ -137,6 +137,14 @@ def to_bytes(s):
     assert m.string_length(u"💩".encode("utf8")) == 4
 
 
+def test_bytearray_to_string():
+    """Tests the ability to pass bytearray to C++ string-accepting functions.  Note that this is
+    one-way: the only way to return bytes to Python is via the pybind11::bytearray class."""
+    # Issue #2799
+    assert m.string_length(bytearray(b"Hi")) == 2
+    assert m.strlen(bytearray(b"bytearray")) == 9
+
```

Comment:
Could you please add tests with empty `bytearray`s?

And with malformed utf-8, to prove that there really isn't any encoding (see e.g. `malformed_utf8` in test_pytypes.py).

I looked at the Python C code underneath `PyByteArray_AsString`, it special-cases empty arrays. But in any case, tests for corner cases are best practice.

Reply:
Sure, an empty bytearray test case will be added.
For malformed_utf8, I'm not sure it's the same case as casting to python str. In test_pytypes, it's casting a single byte to str, and str is not fulfilled with a utf-8 resulting "b'\\x80'". While passing b"\x80" to C++ string it would be treated as a byte and has size 1. Is this the case you're looking for?

Reply:
The motivation/idea is very simple and high-level:

* We want to be sure this works for malformed utf-8.

We know the current implementation does. We want to be sure that's not accidentally broken somehow (e.g. refactoring).

My suggestion was based on past experience, where that actually happened.

Reply:
I see your point, the test is added. I'm trying to make sure `assert m.string_length(bytearray(b"\x80")) == 1` is the behavior we're looking for :blush:

Reply:
Wouldn't it just be better to parameterize the bytes_to_str tests and have them be rerun with all the bytes args wrapped as bytearrays(via pytest fixture for instance? or by having a function that either returns the bytes or the bytes wrapped in a bytearray depending on test parameters)

Reply:
Hi @Skylion007, I also came up wrapping up the params for better coverage. Considering there're at least 3 dimensions (bytearray vs bytes, string_lenth vs strlen, bytearray with/without params), and a few corner cases(empty bytearray), I would suggest the current intuitive version should give us similar coverage with . What do you think?

Reply:
Sounds good.